Examples/Algorithms/Digitization/src/ModuleClusters.cpp

```diff
@@ -0,0 +1,326 @@
+            // dimension loop
+            matched = false;
+            break;
+          }
+        }  // Loop over `k' (non-geo dimensions)
+        if (matched) {
+          // The value under consideration matched at least one
+          // associated to the current cluster so no need to keep
+          // checking others in current cluster
+          break;
+        }
+      }  // Loop on current cluster
+      if (matched) {
+        // Claim value `j'
+        used.at(j) = true;
+        thisvec.push_back(std::move(values.at(j)));
+      }
+    }  // Loop on `j'
+  }    // Loop on `i'
+  return retv;
+}
+
+ModuleClusters::ModuleValue ModuleClusters::squash(
+    std::vector<ModuleClusters::ModuleValue>& values) {
+  ModuleValue mval;
+  Acts::ActsScalar tot = 0;
+  Acts::ActsScalar tot2 = 0;
+  std::vector<Acts::ActsScalar> weights;
+
+  // First, start by computing cell weights
+  for (ModuleValue& other : values) {
+    if (std::holds_alternative<Cluster::Cell>(other.value)) {
+      weights.push_back(std::get<Cluster::Cell>(other.value).activation);
+    } else {
+      weights.push_back(1);
+    }
+    tot += weights.back();
+    tot2 += weights.back() * weights.back();
+  }
+
+  // Now, go over the non-geometric indices
+  for (size_t i = 0; i < values.size(); i++) {
+    ModuleValue& other = values.at(i);
+    for (size_t j = 0; j < other.paramIndices.size(); j++) {
+      auto idx = other.paramIndices.at(j);
+      if (std::find(m_geoIndices.begin(), m_geoIndices.end(), idx) ==
+          m_geoIndices.end()) {
+        mval.paramIndices.push_back(idx);
+        if (mval.paramValues.size() < j) {
+          mval.paramValues.push_back(0);
+          mval.paramVariances.push_back(0);
+        }
+        Acts::ActsScalar f = weights.at(i) / (tot > 0 ? tot : 1);
+        Acts::ActsScalar f2 =
+            weights.at(i) * weights.at(i) / (tot2 > 0 ? tot2 : 1);
+        mval.paramValues.at(j) += f * other.paramValues.at(j);
+        mval.paramVariances.at(j) += f2 * other.paramVariances.at(j);
+      }
+    }
+  }
+
+  // Now do the geometric indices
+  Cluster clus;
+
+  const auto& binningData = m_segmentation.binningData();
+  Acts::Vector2 pos(0., 0.);
+  Acts::Vector2 var(0., 0.);
+
+  size_t b0min = SIZE_MAX;
+  size_t b0max = 0;
+  size_t b1min = SIZE_MAX;
+  size_t b1max = 0;
+
+  for (size_t i = 0; i < values.size(); i++) {
+    ModuleValue& other = values.at(i);
+    if (not std::holds_alternative<Cluster::Cell>(other.value)) {
+      continue;
+    }
+
+    Cluster::Cell ch = std::get<Cluster::Cell>(other.value);
+    auto bin = ch.bin;
+
+    size_t b0 = bin[0];
+    size_t b1 = bin[1];
+
+    b0min = std::min(b0min, b0);
+    b0max = std::max(b0max, b0);
+    b1min = std::min(b1min, b1);
+    b1max = std::max(b1max, b1);
+
+    float p0 = binningData[0].center(b0);
+    float w0 = binningData[0].width(b0);
+    float p1 = binningData[1].center(b1);
+    float w1 = binningData[1].width(b1);
```

Comment:
With taking the with of the bin as weight of the merging step, you assume that the particles have very similar paths through the pixel (and travsere it fully), or do I misunderstand this?

Reply:
The widths are only used to compute the variance of the prediction, assuming a uniform distribution. The "weight" of each cell is still it's charge ("activation"). I did it this way because I wanted to avoid using the variance generator to keep this step separated from the digitizer.

Reply:
As we discussed the other day this way of computing the error overestimates the appropriate variance. I think we can leave it as such for now, and I'll just add a comment saying so.